### PR TITLE
Revamp the contributor guide 

### PR DESCRIPTION
**Pull request by greg0ire**
While some sections feel redundant, other sections were missing, like
the one about choosing the right branch.

This only addresses the first question of #472 

@mpdude please review.

### CODE REVIEW

File: source/contribute.rst [hunk 1]
```diff
@@ -41,128 +41,150 @@ contributing back to a Doctrine project.
...
 
-Primary Branches
-----------------
+If you want to deprecate something, you should target the next minor
+branch: upgrading to a new patch version should not result in any
+additional deprecations.
 
-The **doctrine** repository holds the following primary branches:
+If you want to add a new feature, you should target the next minor
+branch, because it might have side effects that break existing code.
```
Comment (SenseException) [inline source/contribute.rst:124]:
```suggestion
If you want to add a new feature, you should target the next minor
branch, because it might have unintentional side effects that break existing code.
```

**Review by SenseException — APPROVED** [review]
I'd like to see more reviewers for this.

**Comment by mpdude** [comment]
I‘ll have to review the branch selection rules later, with a clear mind.

In general, I'd say there is a lot of information in this chapter that should be routine for people familiar with `git`. 

Do we really need to provide all that? Do we anticipate contributions from users who haven’t done all that before?

**Comment by greg0ire** [comment]
> Do we really need to provide all that? Do we anticipate contributions from users who haven’t done all that before?

There's indeed pretty basic stuff, so we could remove more than I already did. That being said, here are the most common issues I see:

- people who develop directly on 2.3.x instead of creating a topic branch, and are then are asked to retarget on 2.4.x
- people who don't know they can force push and instead close their PR and open a new one when asked to retarget their branch
- bad commit messages (typically `fix for #1234`, where they don't explain what they did and how that's supposed to fix anything) 
- good commit message body, and commit message subject as suggested by github, so typically "Update path/to/file.php"
- non-atomic commits, with unexplained changed mixed with changes relating to the main topic of the pull request
- PR that passes the build, but with commits that don't pass the build individually
- commits that don't bring extra understanding of what's being done but rather show the struggle ("fix cs")

**Comment by greg0ire** [comment]
ping @mpdude :slightly_smiling_face:

File: source/contribute.rst [hunk 2]
```diff
@@ -41,128 +41,150 @@ contributing back to a Doctrine project.
...
 .. code-block:: console
 
     $ cd dbal
-    $ git remote add doctrine git://github.com/doctrine/dbal.git
+    $ git remote add fork git@github.com:username/dbal.git
 
-Branching from the default branch
----------------------------------
+You should now have two remotes: one called ``doctrine``, another called
+``fork``.
```
Comment (mpdude) [inline source/contribute.rst:66]: This should be familiar to anyone who has contributed on GitHub. For first time contributors, there is probably enough material out there on how to get started, including GitHub.com documentation. 

IMHO we do not need to write/maintain this information ourselves.

File: source/contribute.rst [hunk 3]
```diff
@@ -41,128 +41,150 @@ contributing back to a Doctrine project.
...
-The default branch is the branch you see when you enter the repository page on GitHub.
+Choosing the right branch
+-------------------------
 
-.. image:: ../images/default-branch.png
-   :alt: The default branch
-   :style: margin-bottom: 20px
+Before you start working on a feature or a bug fix, you need to choose
+which branch you are going to contribute to. We try hard to follow
+semver, which means there are three choices.
```
Comment (mpdude) [inline source/contribute.rst:73]: Does that apply to all Doctrine projects?

File: source/contribute.rst (hunk 2 above)
Comment (greg0ire) [reply source/contribute.rst:66]: Are you suggesting I remove this first paragraph, or replace it with a link? If the latter which one? This, maybe: https://docs.github.com/en/get-started/quickstart/github-flow ?

**Comment by mpdude** [comment]
Still have to go through the textual changes. 

Regarding the other points you brought up:

When we ask people to re-target PRs, can we expect/assume they find these instructions?

Would it help if we set up a canned reply in GitHub, so that when we ask people to change the target branch we could include a pointer to this documentation?

Regarding bad commit messages, meaningless commits etc.: Would it make sense to discuss squash-merging commits like Symfony does? That would help to hide “how the sausage is being made” and have a clean history with one commit per change (= PR).

File: source/contribute.rst (hunk 2 above)
Comment (mpdude) [reply source/contribute.rst:66]: I'd say https://docs.github.com/en/get-started/quickstart/contributing-to-projects explains how to fork a repo, clone the fork, create a branch, push it and make a PR.

https://docs.github.com/en/get-started/quickstart/hello-world is even more basic and goes through GitHub fundamentals.

In general, it's really hard to tell what would help and invite people with little to no experience to contribute, when you've been using Git or GitHub for a long time. I tend to forget about how it was to learn.

File: source/contribute.rst [hunk 4]
```diff
@@ -41,128 +41,150 @@ contributing back to a Doctrine project.
...
 
-In this DBAL example, it's the branch with the name **2.11.x**. The branch name reflects the
-current lowest supported version of a repository.
+Let's say the latest version of the project is 2.3.1. The following
+branches should exist and are the possible choices:
```
Comment (mpdude) [inline source/contribute.rst:76]: What would be the best way to find out for a given Doctrine project what the current stable/bugfix/... version is?

Along the lines of...
--
You can find out at https://www.doctrine-project.org/projects/orm.html [<-- applies to ORM only] what the current stable as well as the upcoming minor and major versions are.

For example, let's say the current stable version is 2.3.1. This makes 2.4 as well as 3.0 the next minor and major version releases. In the repository, there will be corresponding branches:

Comment (greg0ire) [reply source/contribute.rst:76]: We sometimes forget to update the project metadata, so the best way is really to find what the latest (or rather, highest) version number is, and derive the branch information from that.

File: source/contribute.rst [hunk 5]
```diff
@@ -41,128 +41,150 @@ contributing back to a Doctrine project.
...
+The constraints we have are:
 
-.. code-block:: console
+- Every patch release should be more stable than the previous one.
+- An upgrade to a new minor version should not require any change on the
+  user side.
+- There should be a clear upgrade path from one major version to the
+  next.
+- We want to keep the branches as similar as possible to ease
+  maintenance.
```
Comment (mpdude) [inline source/contribute.rst:94]: The general rules are as follows:
* Patch releases should be as stable as possible. They fix bugs, but in avoid making changes as far as possible.
* Improvements and additions are being made on the upcoming minor release branch. 
* An upgrade to a new minor version must not require any changes on the user's side.
* There must be a clear upgrade path from one major version to the next. That means necessary changes on the user's side will be advertised through deprecation notices, if anyway possible.
* Deprecations may be added for new minor versions only.
* Before a feature is removed (or an API changed) in a major version release, a corresponding deprecation notice must be included in a minor version release preceding it. 
* If a replacement feature and/or alternative API will be provided, it must be available in the minor release that adds the deprecation for the old feature.

And maybe one disputable extra item?

* When a minor release brings (major) changes in behavior, ideally those would be opt-in to avoid surprises during the upgrade process.

Comment (greg0ire) [reply source/contribute.rst]: > When a minor release brings (major) changes in behavior, ideally those would be opt-in to avoid surprises during the upgrade process.

That sounds like a breaking change, and wouldn't go in a minor release :thinking: 

I don't really understand what you mean I'm afraid.

Comment (mpdude) [reply source/contribute.rst]: I notice that this is not editing the contributor's documentation, but instead making policies for the Doctrine project 🥶?

File: source/contribute.rst [hunk 6]
```diff
@@ -41,128 +41,146 @@ contributing back to a Doctrine project.
...
+behavior of the code, you should still target the next minor branch
+unless the refactoring is necessary for a bugfix. Let us not take any
+unnecessary risks.
+
+If you want to deprecate something, you should target the next minor
+branch: upgrading to a new patch version should not result in any
+additional deprecations.
+
+If you want to add a new feature, you should target the next minor
+branch, because it might have unintentional side effects that break existing code.
```
Comment (mpdude) [inline source/contribute.rst:120]: The next minor version branch may include:
* Refactorings, unless they are necessary for bugfixes. This is to avoid unnecessary risks.
* Adding deprecations – remember to update the `UPGRADE.md` file as well!
* Adding new features and/or public APIs

File: source/contribute.rst [hunk 7]
```diff
@@ -41,128 +41,146 @@ contributing back to a Doctrine project.
...
+
+If you want to remove or change a public API, you should target the next
+major branch. If you change a public API, you should provide an upgrade
+path for the end user. Ideally, you should make the new API available on
+the next minor branch, deprecate the old API, and remove it in the next
+major.
```
Comment (mpdude) [inline source/contribute.rst:126]: On the next major version branch, usually only deprecations notices, deprecated features and necessary compatibility layers are being removed.

Avoid suprising changes in public APIs on the next major version branch if anyhow possible. In order to provide a smooth upgrade path to users, the preferred solution is to add deprecations and possibly alternative APIs in a preceding minor version, and only remove the deprecations and compatibility layers in the next major release.

File: source/contribute.rst [hunk 8]
```diff
@@ -41,128 +41,146 @@ contributing back to a Doctrine project.
...
-
-Newly introduced changes to 2.11.x will be up-merged at a later point in time
-to newer version branches (e.g. 2.12.x, 3.0.x). This way you don't have to
-re-introduce a new fix or feature of 2.11.x with another pull request to
-the other version branches.
-
-Keeping the default branch up-to-date!
---------------------------------------
+Creating a topic branch
+-----------------------
```
Comment (mpdude) [inline source/contribute.rst:131]: Same as before – do we really need this entire section?

File: source/contribute.rst (hunk 6 above)
Comment (mpdude) [reply source/contribute.rst:120]: What about changing minimum dependency requirements?

File: source/contribute.rst (hunk 5 above)
Comment (greg0ire) [reply source/contribute.rst]: That makes a lot more sense, yeah, let me add something about it.

Comment (greg0ire) [reply source/contribute.rst]: > I notice that this is not editing the contributor's documentation, but instead making policies for the Doctrine project cold_face?

Do you mean this feels out of place? Because https://www.doctrine-project.org/policies.html could be a good place for this.

File: source/contribute.rst (hunk 4 above)
Comment (greg0ire) [reply source/contribute.rst:58]: There isn't AFAIK. I should create another PR to https://www.doctrine-project.org/contribute/maintainer/index.html I suppose.

File: source/contribute.rst (hunk 6 above)
Comment (greg0ire) [reply source/contribute.rst:120]: Changing minimum dependency requirements is OK on a patch branch if it's absolutely necessary to fix a bug. I don't think that's often the case, and I think we don't have to mention this here.

File: source/contribute.rst (hunk 8 above)
Comment (greg0ire) [reply source/contribute.rst:131]: For that one I know the answer, and that's yes. There are way too many contributors contributing directly to e.g. 2.14.x directly.

File: source/contribute.rst [hunk 9]
```diff
@@ -270,73 +313,20 @@ with both the code and the docs here.
...
 
-    $ git push origin :fix-weird-bug
+1. Fetch all new commits: `git fetch doctrine``.
+2. Rebase on what you fetched interactively:
+   ``git rebase --interactive doctrine/2.3.x # use -i for short``
+3. A window will show up with many lines, remove every line but the last
+   ones, which should correspond to your commits.
+4. If you run into a conflict, fix it with ``git mergetool``, then
+   continue on your merry way with ``git rebase --continue``.
+5. Force push to overwrite the previous version : ``git push --force``.
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply source/contribute.rst:329]: The paragraph is "typically because the build had an issue", and I personally don't approve pull requests that have a red build.

File: source/contribute.rst (hunk 5 above)
Comment (mpdude) [reply source/contribute.rst]: I mean, who am I to write or direct rules that mandate how the Doctrine shall organize/maintain its code 😐

Comment (greg0ire) [reply source/contribute.rst]: Dunno… a future maintainer hopefully :crossed_fingers: ?

File: source/contribute.rst [hunk 10]
```diff
@@ -41,128 +41,144 @@ contributing back to a Doctrine project.
...
+  through deprecation notices, if anyway possible.
+- Deprecations may be added for new minor versions only.
+- Before a feature is removed (or an API changed) in a major version
+  release, a corresponding deprecation notice must be included in a
+  minor version release preceding it.
+- If a replacement feature and/or alternative API will be provided, it
+  must be available in the minor release that adds the deprecation for
+  the old feature.
+
+With that in mind:
```
Comment (mpdude) [inline source/contribute.rst:92]: Here we have two sentences that end with a colon

File: source/contribute.rst [hunk 11]
```diff
@@ -41,128 +41,144 @@ contributing back to a Doctrine project.
...
-Keeping the default branch up-to-date!
---------------------------------------
+Creating a topic branch
+-----------------------
 
-Once all this is done, you'll be able to keep your local branches up to
-date with the following command:
+Time might have elapsed since the last time you contributed or since you
+cloned the repository. You might want to fetch the latest changes from
+the ``doctrine`` remote:
```
Comment (mpdude) [inline source/contribute.rst:133]: Maybe it’s because we removed an initial section, but nothing in the above text supports the assumption that the remote might be called like this

File: source/contribute.rst [hunk 12]
```diff
@@ -41,128 +41,144 @@ contributing back to a Doctrine project.
...
 .. code-block:: console
 
     $ git fetch doctrine
 
-Branching Model
----------------
-
-The following names will be used to differentiate between the different
-repositories:
+I know it sounds awful, but the next step is to deal with one of the 2
```
Comment (mpdude) [inline source/contribute.rst:139]: AFAIK these two are cache invalidation, naming things and off-by-one bugs

File: source/contribute.rst [hunk 13]
```diff
@@ -41,128 +41,144 @@ contributing back to a Doctrine project.
...
--  **doctrine/\*** Maintenance branches of existing releases.
-
-These branches exist in parallel and are defined as follows:
+.. code-block:: console
 
-**doctrine/2.11.x** is the branch where the source code of **HEAD**
-always reflects the latest version. Each released stable version will be
-a tagged commit in a **doctrine/\*** branch. Each released unstable
-version will be a tagged commit in the **doctrine/2.11.x** branch.
+    $ git switch --create your-sql-support doctrine/2.4.x # use -c for short
```
Comment (mpdude) [inline source/contribute.rst:149]: Makes the same assumption

File: source/contribute.rst (hunk 8 above)
Comment (mpdude) [reply source/contribute.rst:129]: Then, let's explain it...

You should avoid making changes directly on the patch, minor or major release branches, also in your fork of the repository. That makes it difficult for you to have more than one pull request at a time, and also complicates matters should you need to rebase your work to another branch some time later on. Instead, for every change that you would like to propose, create a dedicated topic branch. 

Note: If that does not sound all too familiar, you might want to read about GitHub Flow for additional background information https://docs.github.com/en/get-started/quickstart/github-flow.

File: source/contribute.rst [hunk 14]
```diff
@@ -41,128 +41,144 @@ contributing back to a Doctrine project.
...
 
-Once all this is done, you'll be able to keep your local branches up to
-date with the following command:
+Time might have elapsed since the last time you contributed or since you
+cloned the repository. You might want to fetch the latest changes from
+the ``doctrine`` remote:
 
 .. code-block:: console
 
     $ git fetch doctrine
```
Comment (mpdude) [inline source/contribute.rst:137]: Let's move that a bit down

File: source/contribute.rst [hunk 15]
```diff
@@ -198,6 +214,31 @@ this in case you want to read more about this:
...
+
+.. code-block:: console
+
+    $ git push --set-upstream fork your-sql-support # use -u for short
+
+Creating the pull request
+-------------------------
+
+New pull requests are created with the repository's default branch as
+base branch, and that might not be what you want. Make sure to pick the
```
Comment (mpdude) [inline source/contribute.rst:227]: The base branch should be what you chose according to the guidelines mentioned in the beginning of this chapter. It is also the branch name that you provided when you created your topic branch.

File: source/contribute.rst [hunk 16]
```diff
@@ -198,6 +214,31 @@ this in case you want to read more about this:
...
+correct branch when creating the pull request. If you do not, it can
+still be changed afterwards.
```
Comment (mpdude) [inline source/contribute.rst:229]: That'a a bit hidden – on GitHub, you have to click "Edit" next to the pull request title to change the title and base branch

File: source/contribute.rst [hunk 17]
```diff
@@ -270,73 +311,20 @@ with both the code and the docs here.
...
-Once your work has been merged by the branch maintainer, it will no
-longer be necessary to keep the local branch or remote branch, so you
-can remove them!
-
-Sync your local 2.11.x branch:
-
-.. code-block:: console
+Sometimes, you will need to rebase your branch on the latest changes,
+typically because the build had an issue unrelated to your changes, and
+that issue has been fixed after you created your branch.
```
Comment (mpdude) [inline source/contribute.rst:316]: Basically, a ebase takes all changes on your topic branch and moves them to another starting point. This starting point was the Doctrine branch that you chose when you created your topic branch, at that point in time. The rebase will move your changes to the current state of this branch.

File: source/contribute.rst [hunk 18]
```diff
@@ -270,73 +311,20 @@ with both the code and the docs here.
...
-
-.. code-block:: console
+Here is how to proceed if you need to rebase on ``2.3.x``:
 
-    $ git push origin :fix-weird-bug
+1. Fetch all new commits: `git fetch doctrine``.
+2. Rebase on what you fetched interactively:
+   ``git rebase --interactive doctrine/2.3.x # use -i for short``
+3. A window will show up with many lines, remove every line but the last
+   ones, which should correspond to your commits.
```
Comment (mpdude) [inline source/contribute.rst:324]: Hm, that's pretty vague, and every line removed potentially is a lost commit. A nice footgun, eh?

As long as you're only rebasing on the same branch that you initially started out with, does it have to be an interactive rebase at all?

**Review by mpdude — COMMENTED** [review]
IMHO this is already in a very good shape, keep going!

Added a next round of feedback – sorry I missed some of the points last time.

File: source/contribute.rst (hunk 18 above)
Comment (mpdude) [reply source/contribute.rst:324]: (Assuming we're writing all this for people rather new to git and contributing on GitHub, so we should not talk them into complicated situations where they might lose work and/or get frustrated)

**Comment by mpdude** [comment]
Regarding https://github.com/doctrine/doctrine-website/pull/473#issuecomment-1403716535

> * bad commit messages (typically fix for #1234, where they don't explain what they did and how that's supposed to fix anything)
> * good commit message body, and commit message subject as suggested by github, so typically "Update path/to/file.php"
> * non-atomic commits, with unexplained changed mixed with changes relating to the main topic of the pull request
> * PR that passes the build, but with commits that don't pass the build individually
> * commits that don't bring extra understanding of what's being done but rather show the struggle ("fix cs")

IMHO squash-merging PRs could solve all those problems. We could also ask contributors to keep the initial PR description an up-to-date description and summary of the change and use that as the commit message for the merge. Symfony does it that way as well.

Where and how could this best be discussed with the relevant decision-makers from the Doctrine core team?

File: source/contribute.rst (hunk 14 above)
Comment (greg0ire) [reply source/contribute.rst:137]: Sorry, what exactly would you like me to move where, and for what reason?

Comment (greg0ire) [reply source/contribute.rst:137]: Disagree, see https://github.com/doctrine/doctrine-website/pull/473#discussion_r1130629469

File: source/contribute.rst (hunk 18 above)
Comment (mpdude) [reply source/contribute.rst]: Cool

Comment (mpdude) [reply source/contribute.rst]: Let’s have a dedicated section about changing the target branch and the rebase operations that are necessary for this. 

Additionally, prepare a canned reply that we can use in PRs whenever we ask people to change the target branch. This should point to that documentation, mention the correct branch and tell people that they _need not_ close the PR but can update it.

File: source/contribute.rst (hunk 12 above)
Comment (mpdude) [reply source/contribute.rst:169]: To be honest, the name chosen does not really matter much

File: source/contribute.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (greg0ire) [reply source/contribute.rst:179]: Ok so I think I addressed that.

File: source/contribute.rst (hunk 18 above)
Comment (greg0ire) [reply source/contribute.rst]: Addressed.

File: source/contribute.rst [hunk 19]
```diff
@@ -41,128 +41,144 @@ contributing back to a Doctrine project.
...
--  Setup a `GitHub <https://github.com>`_ account.
--  Fork the repository of the project you want to contribute to. In this example
-   it will be `DBAL <https://github.com/doctrine/dbal>`_
--  Clone your fork locally
+We are assuming you have cloned the project, have a Github account, and
+have forked the project. If any of this sounds unfamiliar, please `learn
+about that
+<https://docs.github.com/en/get-started/quickstart/contributing-to-projects>`_
+first.
+
```
[the remark this answers is not in the document]
Comment (mpdude) [reply source/contribute.rst:49]: > Or to rename it… should be less confusing.

I think I misunderstood this, I thought you wanted to rename `origin` to something else.

I agree, there are people who clone the original repo and add their fork as a second remote, or they clone their fork and add `upstream` or similar.

Either way, wouldn't it be safe if we suggest to add a `doctrine` remote pointing to the original Doctrine project's repo, and use that remote name in all other git commands?

[the remark this answers is not in the document]
Comment (mpdude) [reply source/contribute.rst:49]: I just see that in fact, you do that – rename `origin` to `doctrine`. That may be wrong, depending on how people set up their clone. And if they follow the GitHub instruction linked above, it will definetly be wrong, since those instructions end up with `origin`= the fork.

File: source/contribute.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (mpdude) [reply source/contribute.rst:179]: Great

File: source/contribute.rst [hunk 20]
```diff
@@ -41,128 +41,177 @@ contributing back to a Doctrine project.
...
+    $ git branch --delete 2.4.x # use -d for short
+
+Later, if you want to see how things are on ``doctrine/2.4.x``, you can
+use ``--detach``:
+
+.. code-block:: console
+
+    $ git switch --detach doctrine/2.4.x # use -d for short
+
```
Comment (mpdude) [inline source/contribute.rst:197]: Starting from "In fact..." – do we really want to suggest this, and is it relevant in the context of creating a feature branch and committing on this?

Yes, it's an additional safeguard, but nothing you need to repeat every time you create a new branch.

File: source/contribute.rst [hunk 21]
```diff
@@ -198,6 +247,37 @@ this in case you want to read more about this:
 - `Commit message style for git <https://commit.style/>`_
 - `A note about git commit messages <https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html>`_
 
+Once you are done, you can push your branch to your fork:
+
+.. code-block:: console
+
+    $ git push --set-upstream fork your-sql-support # use -u for short
```
Comment (mpdude) [inline source/contribute.rst:254]: `fork`? 🤔

File: source/contribute.rst [hunk 22]
```diff
@@ -198,6 +247,37 @@ this in case you want to read more about this:
...
+The default branch is the branch you see when you enter the repository
+page on GitHub.
+
+.. image:: ../images/default-branch.png
+   :alt: The default branch
+   :style: margin-bottom: 20px
+
+In this DBAL example, it's the branch with the name **2.11.x**. The
+branch name reflects the current lowest supported version of a
+repository.
```
Comment (mpdude) [inline source/contribute.rst:279]: Relevant information?

File: source/contribute.rst [hunk 23]
```diff
@@ -267,76 +347,67 @@ sense to revert a commit documenting a feature without also reverting
...
-Topic Branch Cleanup
---------------------
-
-Once your work has been merged by the branch maintainer, it will no
-longer be necessary to keep the local branch or remote branch, so you
-can remove them!
+Here is how to proceed if you need to rebase on ``2.3.x``:
 
-Sync your local 2.11.x branch:
+1. Fetch all new commits: `git fetch doctrine``.
```
Comment (mpdude) [inline source/contribute.rst:368]: 0. Switch to the branch you'd like to rebase.

File: source/contribute.rst [hunk 24]
```diff
@@ -267,76 +347,67 @@ sense to revert a commit documenting a feature without also reverting
...
 
-Remove your local topic branch using -d to ensure that it has been merged by
-upstream. Branch -d will not delete a branch that is not an ancestor of
-your current head.
+Another case where you need a rebase is when you want to change the
+target branch of your PR. For instance, you might have created your PR
+against ``2.3.x`` but you are told to change it to ``2.4.x``. In that
+case, if ``2.4.x`` does not have all the commits that ``2.3.x`` has, you
+need to drop these commits while rebasing: it is not your responsibility
+to merge them up.
```
Comment (mpdude) [inline source/contribute.rst:385]: Good point, I was not aware of that!

Comment (mpdude) [reply source/contribute.rst:385]: Wait, subconsciously I was.

```
git checkout my-topic
git rebase doctrine/2.3.x --onto doctrine/2.4.x
```

No need to pick the right commits. Am I missing something?

File: source/contribute.rst [hunk 25]
```diff
@@ -267,76 +347,67 @@ sense to revert a commit documenting a feature without also reverting
...
-Next, merge or rebase your commit against **doctrine/2.11.x**. With your
-work done in a **local** topic branch, you'll want to assist upstream
-merge by rebasing your commits. You can either do this manually with
-``fetch`` then ``rebase``, or use the ``pull --rebase`` shortcut. You
-may encounter merge conflicts, which you should fix and then mark as
-fixed with ``add``, and then continue rebasing with
-``rebase --continue``. At any stage, you can abort the rebase with
-``rebase --abort`` unlike nasty merges which will leave files strewn
-everywhere.
+On upstream changes
```
Comment (mpdude) [inline source/contribute.rst:353]: "on" or "onto"? Not a native speaker myself. Git CLI suggests `onto`.

File: source/contribute.rst [hunk 26]
```diff
@@ -267,76 +347,67 @@ sense to revert a commit documenting a feature without also reverting
...
-        Delete a branch irrespective of its merged status.
+1. Fetch all new commits: `git fetch doctrine``.
+2. Rebase on what you fetched:
+   ``git rebase --interactive doctrine/2.3.4 # use -i for short``
+3. You will be presented with a list of commits. You need to drop all
+   commits that are not yours. You can do that by replacing ``pick`` with
+   ``drop``.
+4. If you run into a conflict, fix it with ``git mergetool``, then
+   continue on your merry way with ``git rebase --continue``.
+5. Force push to overwrite the previous version : ``git push --force``.
```
Comment (mpdude) [inline source/contribute.rst:403]: 6. Update the Pull Request target in GitHub

**Comment by mpdude** [comment]
Sorry for all the nitpicking, I owe you a beer once we happen to meet IRL

File: source/contribute.rst (hunk 20 above)
Comment (greg0ire) [reply source/contribute.rst:197]: Let's remove it, it makes the guide longer.

File: source/contribute.rst (hunk 22 above)
Comment (greg0ire) [reply source/contribute.rst:279]: Are you asking if this sentence is relevant, or do you want me to add relevant information?

File: source/contribute.rst [hunk 27]
```diff
@@ -267,76 +347,67 @@ sense to revert a commit documenting a feature without also reverting
...
-Once your work has been merged by the branch maintainer, it will no
-longer be necessary to keep the local branch or remote branch, so you
-can remove them!
+Here is how to proceed if you need to rebase on ``2.3.x``:
 
-Sync your local 2.11.x branch:
+1. Fetch all new commits: `git fetch doctrine``.
+2. Rebase on what you fetched:
+   ``git rebase doctrine/2.3.x``
+3. If you run into a conflict, fix it with ``git mergetool``, then
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply source/contribute.rst:371]: Yes.

File: source/contribute.rst (hunk 24 above)
Comment (greg0ire) [reply source/contribute.rst:385]: You're probably right, let's just suggest that one method.

File: source/contribute.rst (hunk 25 above)
Comment (greg0ire) [reply source/contribute.rst:353]: From `man git rebase`: "it may not be the best idea to keep rebasing on top of the upstream"

File: source/contribute.rst (hunk 26 above)
Comment (greg0ire) [reply source/contribute.rst:403]: Already mentioned above (and I think it's best to do it before pushing)

**Comment by greg0ire** [comment]
Can't wait for the next FOSSA-EU hackathon!

**Comment by mpdude** [comment]
Oh good ol' times

**Comment by greg0ire** [comment]
@mpdude IMO this is ready, can I get a formal approval from you? @SenseException , would it be sufficient for you or do you wish to get other maintainers to review this? I'd hate to late this rot…

**Comment by SenseException** [comment]
Yes, @mpdude fits perfectly as second reviewer. His approval should make this getting merged. :+1:

File: source/contribute.rst (hunk 19 above)
[the remark this answers is not in the document]
Comment (mpdude) [reply source/contribute.rst:49]: @greg0ire I think this point still needs to be addressed:

Simply renaming `origin` to `doctrine` may not be correct. Depending on how people obtained their clone, `origin` may be their fork or the original (upstream) repo.

When closely following the instructions linked in the previous sentence, then `origin` will be the _fork_, so the rename suggested here is definetly wrong.

File: source/contribute.rst [hunk 28]
```diff
@@ -41,128 +41,170 @@ contributing back to a Doctrine project.
...
+    $ git remote rename origin doctrine
 
--  Enter the dbal directory and add the **doctrine** remote
+Also, please ensure your Github fork is known as ``fork`` locally:
 
 .. code-block:: console
 
-    $ cd dbal
-    $ git remote add doctrine git://github.com/doctrine/dbal.git
+    $ git remote add fork git@github.com:<your-name-here>/orm.git
```
Comment (mpdude) [inline source/contribute.rst:65]: Suggestion for this entire section:

---

To make the example commands given in the rest of this page less ambiguous, we will add two explicit names for remote Git repositories: `doctrine` for the original Doctrine project you're contributing to, and `fork` for your fork of it.

.. code-block:: console
    $ [`cd` into the directory created by `git clone ...`]
    $ git remote add doctrine git@github.com:doctrine/<name-of-project>.git
    $ git remote add fork git@github.com:<your-github-name>/<name-of-project>.git

File: source/contribute.rst [hunk 29]
```diff
@@ -41,128 +41,170 @@ contributing back to a Doctrine project.
...
+.. code-block:: console
+
+    $ git fetch doctrine
 
-First create an appropriately named branch. When starting work on a new
-topic, branch off from **doctrine/2.11.x** or a **doctrine/\*** branch:
+.. note::
+
+   Here, we assume ``doctrine`` points to the upstream repository, and
+   not your fork. If you are not sure, you can check with ``git remote -v``.
```
Comment (mpdude) [inline source/contribute.rst:173]: We don't need this note since we've added the remote name in the beginning section.

File: source/contribute.rst (hunk 21 above)
Comment (mpdude) [reply source/contribute.rst:247]: This is now correct :+1:

File: source/contribute.rst [hunk 30]
```diff
@@ -267,76 +340,56 @@ sense to revert a commit documenting a feature without also reverting
...
+in time. The rebase will move your changes to be based on the current
+state of this branch.
 
-Push your branch to **origin**:
+Here is how to proceed if you need to rebase on ``2.3.x``:
 
-Finished topic branches should be pushed to **origin** for a
-**maintainer** to review and pull into **doctrine** as appropriate:
+1. Switch to the branch you would like to rebase.
+2. Fetch all new commits: `git fetch doctrine``.
```
Comment (mpdude) [inline source/contribute.rst:362]: Is this missing a backtick?

File: source/contribute.rst (hunk 24 above)
Comment (mpdude) [reply source/contribute.rst:380]: When using that method, the "commits you did not author" thingy is not a issue at all, right? If so, we need not mention it.

---

Another case where you need a rebase is when you want to change the
target branch of your PR. For instance, you might have created your PR
against ``2.3.x`` but you are told to change it to ``2.4.x``. In that
case, the following command will pick all changes that you made against the ``2.3.x`` branch, and re-apply them on the current ``2.4.x`` branch.

$ git fetch doctrine
$ git rebase --onto doctrine/2.4.x doctrine/2.3.x your-topic-branch
$ git push --force

**Comment by mpdude** [comment]
@greg0ire Done. Can you easily follow up on the new comments I made? Otherwise, I'll walk you through them one by one.

Not much left – let's push this over the finish line soon!

File: source/contribute.rst (hunk 24 above)
Comment (greg0ire) [reply source/contribute.rst:380]: > When using that method, the "commits you did not author" thingy is not a issue at all, right? If so, we need not mention it.

Not an issue indeed 👍

File: source/contribute.rst (hunk 30 above)
Comment (greg0ire) [reply source/contribute.rst:362]: It is 👍

**Comment by greg0ire** [comment]
>  Can you easily follow up on the new comments I made? Otherwise, I'll walk you through them one by one.

I think I did, but I also think I might have easily missed some comments. Can you mark the resolved ones as such?

File: source/contribute.rst (hunk 2 above)
Comment (mpdude) [reply source/contribute.rst]: ✅

File: source/contribute.rst (hunk 3 above)
Comment (mpdude) [reply source/contribute.rst:76]: ✅

File: source/contribute.rst (hunk 4 above)
Comment (mpdude) [reply source/contribute.rst:79]: ✅

File: source/contribute.rst (hunk 5 above)
Comment (mpdude) [reply source/contribute.rst]: ✅

File: source/contribute.rst (hunk 8 above)
Comment (mpdude) [reply source/contribute.rst:149]: ✅

File: source/contribute.rst (hunk 12 above)
Comment (mpdude) [reply source/contribute.rst:181]: ✅

File: source/contribute.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (mpdude) [reply source/contribute.rst:191]: ✅

**Comment by mpdude** [comment]
For some reason, I cannot mark (everything) as "resolved" 🤔 .

Let's see:

- [ ] https://github.com/doctrine/doctrine-website/pull/473#discussion_r1166559279

File: source/contribute.rst [hunk 31]
```diff
@@ -41,128 +41,175 @@ contributing back to a Doctrine project.
...
-
-Keeping the default branch up-to-date!
---------------------------------------
+    $ git remote remove origin
 
-Once all this is done, you'll be able to keep your local branches up to
-date with the following command:
+To make the example commands given in the rest of this page less
+ambiguous, we will add two explicit names for remote Git repositories:
+`doctrine` for the original Doctrine project you're contributing to, and
```
Comment (mpdude) [inline source/contribute.rst:62]: .rst need double backticks here and in the following lines?

File: source/contribute.rst (hunk 22 above)
Comment (mpdude) [reply source/contribute.rst:277]: ✅

**Review by mpdude — APPROVED** [review]

**Comment by mpdude** [comment]
All comments addressed ✅ .

Awesome work @greg0ire 🥇 and sorry it took so many iterations. I really like the result!

**Comment by greg0ire** [comment]
Thanks for the thorough review!
